<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>workflow • PeakGeneNet</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="workflow">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">PeakGeneNet</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/workflow.html">workflow</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>workflow</h1>
            
      

      <div class="d-none name"><code>workflow.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("pak")</span></span>
<span><span class="co"># pak::pak("weitzela/PeakGeneNet")</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://weitzela.github.io/PeakGeneNet/">PeakGeneNet</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="data-preparation">1. Data Preparation<a class="anchor" aria-label="anchor" href="#data-preparation"></a>
</h2>
<p>Required objects:</p>
<ul>
<li>Gene counts: this is a matrix where the row names should be ensembl
IDs<br><table class="table">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
R18804
</th>
<th style="text-align:right;">
R18805
</th>
<th style="text-align:right;">
R18809
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
ENSRNOG00000000008
</td>
<td style="text-align:right;">
6.226779
</td>
<td style="text-align:right;">
6.150745
</td>
<td style="text-align:right;">
6.388451
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSRNOG00000000082
</td>
<td style="text-align:right;">
7.267295
</td>
<td style="text-align:right;">
7.324605
</td>
<td style="text-align:right;">
7.220238
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSRNOG00000001489
</td>
<td style="text-align:right;">
7.768652
</td>
<td style="text-align:right;">
8.032395
</td>
<td style="text-align:right;">
7.814327
</td>
</tr>
</tbody>
</table>
</li>
<li>Peak counts: a list object with one or more matrices. Matrices
should be in same format as gene counts, with row names as genomic
region where the chromosome names, start, and end locations are
separaated by a punctuation mark (e.g., chr1:123:456 or chr1:123-456).
<table class="table">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
PL119617
</th>
<th style="text-align:right;">
PL119618
</th>
<th style="text-align:right;">
PL119619
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
chr2:103325658:103326019
</td>
<td style="text-align:right;">
6.616630
</td>
<td style="text-align:right;">
6.816892
</td>
<td style="text-align:right;">
6.210107
</td>
</tr>
<tr>
<td style="text-align:left;">
chr2:103337000:103337431
</td>
<td style="text-align:right;">
7.643895
</td>
<td style="text-align:right;">
7.417112
</td>
<td style="text-align:right;">
7.204352
</td>
</tr>
<tr>
<td style="text-align:left;">
chr2:103483778:103484423
</td>
<td style="text-align:right;">
7.449124
</td>
<td style="text-align:right;">
7.336309
</td>
<td style="text-align:right;">
7.031643
</td>
</tr>
<tr>
<td style="text-align:left;">
chr3:78634005:78634640
</td>
<td style="text-align:right;">
6.754136
</td>
<td style="text-align:right;">
6.966854
</td>
<td style="text-align:right;">
7.031643
</td>
</tr>
<tr>
<td style="text-align:left;">
chr3:78652614:78653063
</td>
<td style="text-align:right;">
6.707064
</td>
<td style="text-align:right;">
6.966854
</td>
<td style="text-align:right;">
6.674483
</td>
</tr>
<tr>
<td style="text-align:left;">
chr3:78702369:78702924
</td>
<td style="text-align:right;">
6.890375
</td>
<td style="text-align:right;">
7.033393
</td>
<td style="text-align:right;">
7.272677
</td>
</tr>
</tbody>
</table>
</li>
</ul>
<p>Use these objects to create peak-gene pairs for inclusion in the
correlation analysis. The <code>p2g_info</code> provides information for
individual peaks per gene they will be correlated with. The
<code>correlation_pairs</code> object is used for reference for all of
the correlations that will be run.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p2g_ls</span> <span class="op">=</span> <span class="fu"><a href="../reference/createPeak2GeneObjects.html">createPeak2GeneObjects</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">gene_counts</span><span class="op">)</span>, <span class="va">peak_counts</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map</a></span><span class="op">(</span><span class="va">rownames</span><span class="op">)</span>, </span>
<span>  <span class="fu">biomaRt</span><span class="fu">::</span><span class="fu"><a href="https://huber-group-embl.github.io/biomaRt/reference/useEnsembl.html" class="external-link">useEnsembl</a></span><span class="op">(</span>biomart <span class="op">=</span> <span class="st">"genes"</span>, </span>
<span>                      dataset <span class="op">=</span> <span class="st">"rnorvegicus_gene_ensembl"</span>, version <span class="op">=</span> <span class="fl">109</span><span class="op">)</span>, </span>
<span>  <span class="st">"rn7"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">p2g_ls</span><span class="op">$</span><span class="va">p2g_info</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="op">]</span> <span class="op">|&gt;</span> <span class="fu">kableExtra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr>
<th style="text-align:left;">
unique_id
</th>
<th style="text-align:left;">
region_id
</th>
<th style="text-align:left;">
modality
</th>
<th style="text-align:left;">
ensembl_gene_id
</th>
<th style="text-align:right;">
dist_to_tss
</th>
<th style="text-align:left;">
promoter_peak
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
chr3:78634005:78634640_ATACSeq
</td>
<td style="text-align:left;">
chr3:78634005:78634640
</td>
<td style="text-align:left;">
ATACSeq
</td>
<td style="text-align:left;">
ENSRNOG00000000008
</td>
<td style="text-align:right;">
-977078
</td>
<td style="text-align:left;">
FALSE
</td>
</tr>
<tr>
<td style="text-align:left;">
chr3:78652401:78653178_H3K4me1
</td>
<td style="text-align:left;">
chr3:78652401:78653178
</td>
<td style="text-align:left;">
H3K4me1
</td>
<td style="text-align:left;">
ENSRNOG00000000008
</td>
<td style="text-align:right;">
-958540
</td>
<td style="text-align:left;">
FALSE
</td>
</tr>
<tr>
<td style="text-align:left;">
chr3:78652614:78653063_ATACSeq
</td>
<td style="text-align:left;">
chr3:78652614:78653063
</td>
<td style="text-align:left;">
ATACSeq
</td>
<td style="text-align:left;">
ENSRNOG00000000008
</td>
<td style="text-align:right;">
-958655
</td>
<td style="text-align:left;">
FALSE
</td>
</tr>
<tr>
<td style="text-align:left;">
chr3:78683169:78683832_H3K27ac
</td>
<td style="text-align:left;">
chr3:78683169:78683832
</td>
<td style="text-align:left;">
H3K27ac
</td>
<td style="text-align:left;">
ENSRNOG00000000008
</td>
<td style="text-align:right;">
-927886
</td>
<td style="text-align:left;">
FALSE
</td>
</tr>
<tr>
<td style="text-align:left;">
chr3:78683236:78683697_H3K4me1
</td>
<td style="text-align:left;">
chr3:78683236:78683697
</td>
<td style="text-align:left;">
H3K4me1
</td>
<td style="text-align:left;">
ENSRNOG00000000008
</td>
<td style="text-align:right;">
-928021
</td>
<td style="text-align:left;">
FALSE
</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p2g_ls</span><span class="op">$</span><span class="va">correlation_pairs</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="op">]</span> <span class="op">|&gt;</span> <span class="fu">kableExtra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr>
<th style="text-align:left;">
ensembl_gene_id
</th>
<th style="text-align:left;">
regulatory_element
</th>
<th style="text-align:left;">
target_id
</th>
<th style="text-align:left;">
link_label
</th>
<th style="text-align:left;">
chr
</th>
<th style="text-align:left;">
modality_pair
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
ENSRNOG00000000008
</td>
<td style="text-align:left;">
chr3:78634005:78634640_ATACSeq
</td>
<td style="text-align:left;">
chr3:79610922:79613046_H3K4me3
</td>
<td style="text-align:left;">
distal_peak_to_promoter_peak
</td>
<td style="text-align:left;">
chr3
</td>
<td style="text-align:left;">
ATACSeq-H3K4me3
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSRNOG00000000008
</td>
<td style="text-align:left;">
chr3:78652401:78653178_H3K4me1
</td>
<td style="text-align:left;">
chr3:79610922:79613046_H3K4me3
</td>
<td style="text-align:left;">
distal_peak_to_promoter_peak
</td>
<td style="text-align:left;">
chr3
</td>
<td style="text-align:left;">
H3K4me1-H3K4me3
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSRNOG00000000008
</td>
<td style="text-align:left;">
chr3:78652614:78653063_ATACSeq
</td>
<td style="text-align:left;">
chr3:79610922:79613046_H3K4me3
</td>
<td style="text-align:left;">
distal_peak_to_promoter_peak
</td>
<td style="text-align:left;">
chr3
</td>
<td style="text-align:left;">
ATACSeq-H3K4me3
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSRNOG00000000008
</td>
<td style="text-align:left;">
chr3:78683169:78683832_H3K27ac
</td>
<td style="text-align:left;">
chr3:79610922:79613046_H3K4me3
</td>
<td style="text-align:left;">
distal_peak_to_promoter_peak
</td>
<td style="text-align:left;">
chr3
</td>
<td style="text-align:left;">
H3K27ac-H3K4me3
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSRNOG00000000008
</td>
<td style="text-align:left;">
chr3:78683236:78683697_H3K4me1
</td>
<td style="text-align:left;">
chr3:79610922:79613046_H3K4me3
</td>
<td style="text-align:left;">
distal_peak_to_promoter_peak
</td>
<td style="text-align:left;">
chr3
</td>
<td style="text-align:left;">
H3K4me1-H3K4me3
</td>
</tr>
</tbody>
</table>
<p>See how many relationships will be tested within each category.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p2g_ls</span><span class="op">$</span><span class="va">correlation_pairs</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">ensembl_gene_id</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html" class="external-link">distinct</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">link_label</span>, <span class="va">modality_pair</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">tally</a></span><span class="op">(</span>name <span class="op">=</span> <span class="st">"No. of Correlations"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">janitor</span><span class="fu">::</span><span class="fu"><a href="https://sfirke.github.io/janitor/reference/adorn_totals.html" class="external-link">adorn_totals</a></span><span class="op">(</span>where <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"row"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/across.html" class="external-link">across</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/where.html" class="external-link">where</a></span><span class="op">(</span><span class="va">is.numeric</span><span class="op">)</span>, <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/comma.html" class="external-link">comma</a></span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">kableExtra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr>
<th style="text-align:left;">
link_label
</th>
<th style="text-align:left;">
modality_pair
</th>
<th style="text-align:left;">
No. of Correlations
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
promoter_peak_to_gene
</td>
<td style="text-align:left;">
ATACSeq-RNASeq
</td>
<td style="text-align:left;">
4
</td>
</tr>
<tr>
<td style="text-align:left;">
promoter_peak_to_gene
</td>
<td style="text-align:left;">
H3K27ac-RNASeq
</td>
<td style="text-align:left;">
3
</td>
</tr>
<tr>
<td style="text-align:left;">
promoter_peak_to_gene
</td>
<td style="text-align:left;">
H3K4me3-RNASeq
</td>
<td style="text-align:left;">
4
</td>
</tr>
<tr>
<td style="text-align:left;">
distal_peak_to_gene
</td>
<td style="text-align:left;">
ATACSeq-RNASeq
</td>
<td style="text-align:left;">
292
</td>
</tr>
<tr>
<td style="text-align:left;">
distal_peak_to_gene
</td>
<td style="text-align:left;">
H3K27ac-RNASeq
</td>
<td style="text-align:left;">
121
</td>
</tr>
<tr>
<td style="text-align:left;">
distal_peak_to_gene
</td>
<td style="text-align:left;">
H3K4me1-RNASeq
</td>
<td style="text-align:left;">
302
</td>
</tr>
<tr>
<td style="text-align:left;">
distal_peak_to_gene
</td>
<td style="text-align:left;">
H3K4me3-RNASeq
</td>
<td style="text-align:left;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
distal_peak_to_promoter_peak
</td>
<td style="text-align:left;">
ATACSeq-ATACSeq
</td>
<td style="text-align:left;">
380
</td>
</tr>
<tr>
<td style="text-align:left;">
distal_peak_to_promoter_peak
</td>
<td style="text-align:left;">
ATACSeq-H3K27ac
</td>
<td style="text-align:left;">
410
</td>
</tr>
<tr>
<td style="text-align:left;">
distal_peak_to_promoter_peak
</td>
<td style="text-align:left;">
ATACSeq-H3K4me1
</td>
<td style="text-align:left;">
409
</td>
</tr>
<tr>
<td style="text-align:left;">
distal_peak_to_promoter_peak
</td>
<td style="text-align:left;">
ATACSeq-H3K4me3
</td>
<td style="text-align:left;">
381
</td>
</tr>
<tr>
<td style="text-align:left;">
distal_peak_to_promoter_peak
</td>
<td style="text-align:left;">
H3K27ac-H3K27ac
</td>
<td style="text-align:left;">
120
</td>
</tr>
<tr>
<td style="text-align:left;">
distal_peak_to_promoter_peak
</td>
<td style="text-align:left;">
H3K27ac-H3K4me1
</td>
<td style="text-align:left;">
282
</td>
</tr>
<tr>
<td style="text-align:left;">
distal_peak_to_promoter_peak
</td>
<td style="text-align:left;">
H3K27ac-H3K4me3
</td>
<td style="text-align:left;">
161
</td>
</tr>
<tr>
<td style="text-align:left;">
distal_peak_to_promoter_peak
</td>
<td style="text-align:left;">
H3K4me1-H3K4me3
</td>
<td style="text-align:left;">
409
</td>
</tr>
<tr>
<td style="text-align:left;">
distal_peak_to_promoter_peak
</td>
<td style="text-align:left;">
H3K4me3-H3K4me3
</td>
<td style="text-align:left;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
promoter_peak_to_promoter_peak
</td>
<td style="text-align:left;">
ATACSeq-ATACSeq
</td>
<td style="text-align:left;">
2
</td>
</tr>
<tr>
<td style="text-align:left;">
promoter_peak_to_promoter_peak
</td>
<td style="text-align:left;">
ATACSeq-H3K27ac
</td>
<td style="text-align:left;">
10
</td>
</tr>
<tr>
<td style="text-align:left;">
promoter_peak_to_promoter_peak
</td>
<td style="text-align:left;">
ATACSeq-H3K4me3
</td>
<td style="text-align:left;">
12
</td>
</tr>
<tr>
<td style="text-align:left;">
promoter_peak_to_promoter_peak
</td>
<td style="text-align:left;">
H3K27ac-H3K27ac
</td>
<td style="text-align:left;">
2
</td>
</tr>
<tr>
<td style="text-align:left;">
promoter_peak_to_promoter_peak
</td>
<td style="text-align:left;">
H3K27ac-H3K4me3
</td>
<td style="text-align:left;">
10
</td>
</tr>
<tr>
<td style="text-align:left;">
promoter_peak_to_promoter_peak
</td>
<td style="text-align:left;">
H3K4me3-H3K4me3
</td>
<td style="text-align:left;">
2
</td>
</tr>
<tr>
<td style="text-align:left;">
Total
</td>
<td style="text-align:left;">
<ul><li></ul>
</td>
<td style="text-align:left;">
3,318
</td>
</tr>
</tbody>
</table>
<p>Then, carry out and process the correlations using the following
functions. The count matrices to for used as the input for the
correlation analysis should have samples in rows and features in
columns.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">peak_counts_t</span> <span class="op">=</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map</a></span><span class="op">(</span><span class="va">peak_counts</span>, <span class="kw">function</span><span class="op">(</span><span class="va">.x</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">.x</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">.x</span>, <span class="st">"samp"</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">.x</span><span class="op">)</span>, <span class="st">"samp_id"</span><span class="op">]</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">.x</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="va">count_mat</span> <span class="op">=</span> <span class="fu"><a href="../reference/formatMatrixForCorrelation.html">formatMatrixForCorrelation</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">gene_counts</span><span class="op">)</span>, <span class="va">peak_counts_t</span><span class="op">)</span></span>
<span><span class="va">count_mat</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span> <span class="op">|&gt;</span> <span class="fu">kableExtra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
ENSRNOG00000000008_RNASeq
</th>
<th style="text-align:right;">
ENSRNOG00000000082_RNASeq
</th>
<th style="text-align:right;">
ENSRNOG00000001489_RNASeq
</th>
<th style="text-align:right;">
chr3:78683169:78683832_H3K27ac
</th>
<th style="text-align:right;">
chr3:78708054:78708512_H3K27ac
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
R18804
</td>
<td style="text-align:right;">
6.226779
</td>
<td style="text-align:right;">
7.267295
</td>
<td style="text-align:right;">
7.768652
</td>
<td style="text-align:right;">
5.974274
</td>
<td style="text-align:right;">
5.830707
</td>
</tr>
<tr>
<td style="text-align:left;">
R18805
</td>
<td style="text-align:right;">
6.150745
</td>
<td style="text-align:right;">
7.324605
</td>
<td style="text-align:right;">
8.032395
</td>
<td style="text-align:right;">
6.015659
</td>
<td style="text-align:right;">
5.922595
</td>
</tr>
<tr>
<td style="text-align:left;">
R18809
</td>
<td style="text-align:right;">
6.388451
</td>
<td style="text-align:right;">
7.220238
</td>
<td style="text-align:right;">
7.814327
</td>
<td style="text-align:right;">
6.018896
</td>
<td style="text-align:right;">
5.880515
</td>
</tr>
<tr>
<td style="text-align:left;">
R18810
</td>
<td style="text-align:right;">
6.201142
</td>
<td style="text-align:right;">
7.341687
</td>
<td style="text-align:right;">
7.850113
</td>
<td style="text-align:right;">
6.069272
</td>
<td style="text-align:right;">
6.599047
</td>
</tr>
<tr>
<td style="text-align:left;">
R18819
</td>
<td style="text-align:right;">
6.328846
</td>
<td style="text-align:right;">
7.358075
</td>
<td style="text-align:right;">
7.973985
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
NA
</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">full_cor_res</span> <span class="op">=</span> <span class="fu"><a href="../reference/correlateByChromosome.html">correlateByChromosome</a></span><span class="op">(</span><span class="va">count_mat</span>, <span class="va">p2g_ls</span><span class="op">$</span><span class="va">correlation_pairs</span><span class="op">)</span></span>
<span><span class="co">#&gt; Starting chr1</span></span>
<span><span class="co">#&gt; 0.022 sec elapsed</span></span>
<span><span class="co">#&gt; Starting chr2</span></span>
<span><span class="co">#&gt; 0.007 sec elapsed</span></span>
<span><span class="co">#&gt; Starting chr3</span></span>
<span><span class="co">#&gt; 0.013 sec elapsed</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(regulatory_element, target_id, link_label, chr,</span></span>
<span><span class="co">#&gt; modality_pair, r, n, P)`</span></span>
<span><span class="va">full_cor_res</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">kableExtra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr>
<th style="text-align:left;">
ensembl_gene_id
</th>
<th style="text-align:left;">
regulatory_element
</th>
<th style="text-align:left;">
target_id
</th>
<th style="text-align:left;">
link_label
</th>
<th style="text-align:left;">
chr
</th>
<th style="text-align:left;">
modality_pair
</th>
<th style="text-align:right;">
r
</th>
<th style="text-align:right;">
n
</th>
<th style="text-align:right;">
P
</th>
<th style="text-align:right;">
BH
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
ENSRNOG00000000008
</td>
<td style="text-align:left;">
chr3:78634005:78634640_ATACSeq
</td>
<td style="text-align:left;">
chr3:79610922:79613046_H3K4me3
</td>
<td style="text-align:left;">
distal_peak_to_promoter_peak
</td>
<td style="text-align:left;">
chr3
</td>
<td style="text-align:left;">
ATACSeq-H3K4me3
</td>
<td style="text-align:right;">
0.1656186
</td>
<td style="text-align:right;">
56
</td>
<td style="text-align:right;">
0.2225149
</td>
<td style="text-align:right;">
0.8032780
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSRNOG00000000008
</td>
<td style="text-align:left;">
chr3:78652401:78653178_H3K4me1
</td>
<td style="text-align:left;">
chr3:79610922:79613046_H3K4me3
</td>
<td style="text-align:left;">
distal_peak_to_promoter_peak
</td>
<td style="text-align:left;">
chr3
</td>
<td style="text-align:left;">
H3K4me1-H3K4me3
</td>
<td style="text-align:right;">
0.0949883
</td>
<td style="text-align:right;">
59
</td>
<td style="text-align:right;">
0.4742207
</td>
<td style="text-align:right;">
0.9788157
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSRNOG00000000008
</td>
<td style="text-align:left;">
chr3:78652614:78653063_ATACSeq
</td>
<td style="text-align:left;">
chr3:79610922:79613046_H3K4me3
</td>
<td style="text-align:left;">
distal_peak_to_promoter_peak
</td>
<td style="text-align:left;">
chr3
</td>
<td style="text-align:left;">
ATACSeq-H3K4me3
</td>
<td style="text-align:right;">
-0.0216678
</td>
<td style="text-align:right;">
56
</td>
<td style="text-align:right;">
0.8740562
</td>
<td style="text-align:right;">
0.9817481
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSRNOG00000000008
</td>
<td style="text-align:left;">
chr3:78683169:78683832_H3K27ac
</td>
<td style="text-align:left;">
chr3:79610922:79613046_H3K4me3
</td>
<td style="text-align:left;">
distal_peak_to_promoter_peak
</td>
<td style="text-align:left;">
chr3
</td>
<td style="text-align:left;">
H3K27ac-H3K4me3
</td>
<td style="text-align:right;">
0.0444185
</td>
<td style="text-align:right;">
59
</td>
<td style="text-align:right;">
0.7383421
</td>
<td style="text-align:right;">
0.9916836
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSRNOG00000000008
</td>
<td style="text-align:left;">
chr3:78683236:78683697_H3K4me1
</td>
<td style="text-align:left;">
chr3:79610922:79613046_H3K4me3
</td>
<td style="text-align:left;">
distal_peak_to_promoter_peak
</td>
<td style="text-align:left;">
chr3
</td>
<td style="text-align:left;">
H3K4me1-H3K4me3
</td>
<td style="text-align:right;">
0.1593074
</td>
<td style="text-align:right;">
59
</td>
<td style="text-align:right;">
0.2281274
</td>
<td style="text-align:right;">
0.9237553
</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">processed_cor</span> <span class="op">=</span> <span class="fu"><a href="../reference/processCorrelations.html">processCorrelations</a></span><span class="op">(</span><span class="va">count_mat</span>, <span class="va">full_cor_res</span>, <span class="va">p2g_ls</span><span class="op">$</span><span class="va">correlation_pairs</span>, <span class="va">p2g_ls</span><span class="op">$</span><span class="va">p2g_info</span><span class="op">)</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(ensembl_gene_id, regulatory_element)`</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(ensembl_gene_id, regulatory_element, target_id)`</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(ensembl_gene_id, regulatory_element, target_id)`</span></span>
<span><span class="co">#&gt; ### Correlating ambiguous peaks with other distal peaks assocaited with gene</span></span>
<span><span class="co">#&gt; ### pre fxn: disclude: 74 neg: 22 pos: 22</span></span>
<span><span class="co">#&gt; round 1: disclude: 2 drop: 27 neg: 44 pos: 45</span></span>
<span><span class="co">#&gt; round 2: disclude: 1 drop: 27 neg: 44 pos: 46</span></span>
<span><span class="co">#&gt; round 3: disclude: 1 drop: 27 neg: 44 pos: 46</span></span>
<span><span class="co">#&gt; round 4: disclude: 1 drop: 27 neg: 44 pos: 46</span></span>
<span><span class="co">#&gt; round 5: drop: 28 neg: 44 pos: 46</span></span>
<span><span class="va">processed_cor</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">kableExtra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr>
<th style="text-align:left;">
ensembl_gene_id
</th>
<th style="text-align:left;">
regulatory_element
</th>
<th style="text-align:left;">
link_label
</th>
<th style="text-align:left;">
modality_pair
</th>
<th style="text-align:right;">
r
</th>
<th style="text-align:right;">
BH
</th>
<th style="text-align:right;">
dist_to_tss
</th>
<th style="text-align:left;">
region2gene_dir
</th>
<th style="text-align:right;">
region2gene_conf
</th>
<th style="text-align:left;">
target_id
</th>
<th style="text-align:left;">
chr
</th>
<th style="text-align:right;">
n
</th>
<th style="text-align:right;">
P
</th>
<th style="text-align:left;">
sig_cor
</th>
<th style="text-align:left;">
prom_peak_sig_cor_w_gene
</th>
<th style="text-align:left;">
distal_no_promoter
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
ENSRNOG00000000008
</td>
<td style="text-align:left;">
chr3:78634005:78634640_ATACSeq
</td>
<td style="text-align:left;">
distal_peak_to_promoter_peak
</td>
<td style="text-align:left;">
ATACSeq-ATACSeq
</td>
<td style="text-align:right;">
0.3013569
</td>
<td style="text-align:right;">
0.0043280
</td>
<td style="text-align:right;">
-977078
</td>
<td style="text-align:left;">
pos
</td>
<td style="text-align:right;">
4.001
</td>
<td style="text-align:left;">
chr3:79611428:79612928_ATACSeq
</td>
<td style="text-align:left;">
chr3
</td>
<td style="text-align:right;">
123
</td>
<td style="text-align:right;">
0.0007061
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSRNOG00000000008
</td>
<td style="text-align:left;">
chr3:78702369:78702924_ATACSeq
</td>
<td style="text-align:left;">
distal_peak_to_promoter_peak
</td>
<td style="text-align:left;">
ATACSeq-ATACSeq
</td>
<td style="text-align:right;">
0.3539229
</td>
<td style="text-align:right;">
0.0006360
</td>
<td style="text-align:right;">
-908794
</td>
<td style="text-align:left;">
pos
</td>
<td style="text-align:right;">
4.001
</td>
<td style="text-align:left;">
chr3:79611428:79612928_ATACSeq
</td>
<td style="text-align:left;">
chr3
</td>
<td style="text-align:right;">
123
</td>
<td style="text-align:right;">
0.0000592
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSRNOG00000000008
</td>
<td style="text-align:left;">
chr3:78713352:78714085_ATACSeq
</td>
<td style="text-align:left;">
distal_peak_to_promoter_peak
</td>
<td style="text-align:left;">
ATACSeq-ATACSeq
</td>
<td style="text-align:right;">
0.2480879
</td>
<td style="text-align:right;">
0.0226489
</td>
<td style="text-align:right;">
-897633
</td>
<td style="text-align:left;">
pos
</td>
<td style="text-align:right;">
4.001
</td>
<td style="text-align:left;">
chr3:79611428:79612928_ATACSeq
</td>
<td style="text-align:left;">
chr3
</td>
<td style="text-align:right;">
123
</td>
<td style="text-align:right;">
0.0056622
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSRNOG00000000008
</td>
<td style="text-align:left;">
chr3:78734426:78734839_ATACSeq
</td>
<td style="text-align:left;">
distal_peak_to_promoter_peak
</td>
<td style="text-align:left;">
ATACSeq-ATACSeq
</td>
<td style="text-align:right;">
-0.3866776
</td>
<td style="text-align:right;">
0.0001516
</td>
<td style="text-align:right;">
-876879
</td>
<td style="text-align:left;">
neg
</td>
<td style="text-align:right;">
4.001
</td>
<td style="text-align:left;">
chr3:79611428:79612928_ATACSeq
</td>
<td style="text-align:left;">
chr3
</td>
<td style="text-align:right;">
123
</td>
<td style="text-align:right;">
0.0000100
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSRNOG00000000008
</td>
<td style="text-align:left;">
chr3:78783524:78783997_ATACSeq
</td>
<td style="text-align:left;">
distal_peak_to_promoter_peak
</td>
<td style="text-align:left;">
ATACSeq-ATACSeq
</td>
<td style="text-align:right;">
0.2642814
</td>
<td style="text-align:right;">
0.0135542
</td>
<td style="text-align:right;">
-827721
</td>
<td style="text-align:left;">
pos
</td>
<td style="text-align:right;">
4.001
</td>
<td style="text-align:left;">
chr3:79611428:79612928_ATACSeq
</td>
<td style="text-align:left;">
chr3
</td>
<td style="text-align:right;">
123
</td>
<td style="text-align:right;">
0.0031389
</td>
<td style="text-align:left;">
TRUE
</td>
<td style="text-align:left;">
FALSE
</td>
<td style="text-align:left;">
FALSE
</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.5.2 (2025-10-31)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 24.04.3 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] PeakGeneNet_0.1.0 lubridate_1.9.4   forcats_1.0.1     stringr_1.6.0    </span></span>
<span><span class="co">#&gt;  [5] dplyr_1.1.4       purrr_1.2.1       readr_2.1.6       tidyr_1.3.2      </span></span>
<span><span class="co">#&gt;  [9] tibble_3.3.1      ggplot2_4.0.1     tidyverse_2.0.0  </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] tidyselect_1.2.1     viridisLite_0.4.2    farver_2.1.2        </span></span>
<span><span class="co">#&gt;  [4] blob_1.3.0           filelock_1.0.3       Biostrings_2.78.0   </span></span>
<span><span class="co">#&gt;  [7] S7_0.2.1             fastmap_1.2.0        BiocFileCache_3.0.0 </span></span>
<span><span class="co">#&gt; [10] janitor_2.2.1        digest_0.6.39        timechange_0.3.0    </span></span>
<span><span class="co">#&gt; [13] lifecycle_1.0.5      KEGGREST_1.50.0      RSQLite_2.4.5       </span></span>
<span><span class="co">#&gt; [16] magrittr_2.0.4       compiler_4.5.2       rlang_1.1.7         </span></span>
<span><span class="co">#&gt; [19] sass_0.4.10          progress_1.2.3       tools_4.5.2         </span></span>
<span><span class="co">#&gt; [22] yaml_2.3.12          knitr_1.51           prettyunits_1.2.0   </span></span>
<span><span class="co">#&gt; [25] bit_4.6.0            curl_7.0.0           xml2_1.5.2          </span></span>
<span><span class="co">#&gt; [28] RColorBrewer_1.1-3   withr_3.0.2          BiocGenerics_0.56.0 </span></span>
<span><span class="co">#&gt; [31] desc_1.4.3           grid_4.5.2           stats4_4.5.2        </span></span>
<span><span class="co">#&gt; [34] scales_1.4.0         gtools_3.9.5         biomaRt_2.66.0      </span></span>
<span><span class="co">#&gt; [37] cli_3.6.5            rmarkdown_2.30       crayon_1.5.3        </span></span>
<span><span class="co">#&gt; [40] ragg_1.5.0           generics_0.1.4       rstudioapi_0.18.0   </span></span>
<span><span class="co">#&gt; [43] httr_1.4.7           tzdb_0.5.0           DBI_1.2.3           </span></span>
<span><span class="co">#&gt; [46] cachem_1.1.0         AnnotationDbi_1.72.0 XVector_0.50.0      </span></span>
<span><span class="co">#&gt; [49] matrixStats_1.5.0    vctrs_0.7.0          jsonlite_2.0.0      </span></span>
<span><span class="co">#&gt; [52] IRanges_2.44.0       hms_1.1.4            S4Vectors_0.48.0    </span></span>
<span><span class="co">#&gt; [55] bit64_4.6.0-1        systemfonts_1.3.1    jquerylib_0.1.4     </span></span>
<span><span class="co">#&gt; [58] glue_1.8.0           pkgdown_2.2.0        stringi_1.8.7       </span></span>
<span><span class="co">#&gt; [61] gtable_0.3.6         GenomeInfoDb_1.46.2  GenomicRanges_1.62.1</span></span>
<span><span class="co">#&gt; [64] UCSC.utils_1.6.1     pillar_1.11.1        rappdirs_0.3.4      </span></span>
<span><span class="co">#&gt; [67] htmltools_0.5.9      Seqinfo_1.0.0        R6_2.6.1            </span></span>
<span><span class="co">#&gt; [70] dbplyr_2.5.1         httr2_1.2.2          textshaping_1.0.4   </span></span>
<span><span class="co">#&gt; [73] evaluate_1.0.5       kableExtra_1.4.0     Biobase_2.70.0      </span></span>
<span><span class="co">#&gt; [76] png_0.1-8            tictoc_1.2.1         memoise_2.0.1       </span></span>
<span><span class="co">#&gt; [79] snakecase_0.11.1     bslib_0.9.0          svglite_2.2.2       </span></span>
<span><span class="co">#&gt; [82] xfun_0.56            fs_1.6.6             pkgconfig_2.0.3</span></span></code></pre></div>
</div>
  </main>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Anne Weitzel.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
